src/vision/status_quo/barbara_builds_an_async_executor.md

```diff
@@ -0,0 +1,161 @@
+
+```rust
+pub struct Task {
+    future: *mut (dyn Future<Output = ()> + Send + 'static),
+    ...
+}
+
+unsafe impl Send for Task {}
+unsafe impl Sync for Task {}
+
+pub fn poll_task(task: Arc<Task>) {
+    ...
+    unsafe {
+        Pin::new_unchecked(&mut *task.future).poll(&mut cx);
+    }
+}
+```
+
+Luckily, a colleague of Barbara noticed something wrong. The `wake` method could be called multiple times so multiple copies of the task could exist in the scheduler. The scheduler might not work correctly because of it and a more severe problem was that multiple threads might get copies of the same task from the scheduler and cause a race in polling the future.
+
+Barbara soon got a idea to solve it. She added a state field to the `Task`. By carefully maintaining the state of the task, she could guarantee there are no duplicate tasks in the scheduler and no race can happen when polling the future:
+
+```rust
+const NOTIFIED: u64 = 1;
+const IDLE: u64 = 2;
+const POLLING: u64 = 3;
+const COMPLETED: u64 = 4;
+
+pub struct Task {
+    ...
+    state: AtomicU64,
+}
+
+impl Wake for Task {
+    fn wake(self: Arc<Self>) {
+        let mut state = self.state.load(Relaxed);
+        loop {
+            match state {
+                IDLE => match self
+                    .state
+                    .compare_exchange_weak(IDLE, NOTIFIED, AcqRel, Acquire)
+                {
+                    Ok(_) => self.scheduler.send(self.clone()),
+                    Err(s) => state = s,
+                },
+                POLLING => match self
+                    .state
+                    .compare_exchange_weak(POLLING, NOTIFIED, AcqRel, Acquire)
+                {
+                    Ok(_) => break,
+                    Err(s) => state = s,
+                },
+                _ => break,
+            }
+        }
+    }
+}
+pub fn poll_task(task: Arc<Task>) {
+    let waker = Waker::from(task.clone());
+    let mut cx = Context::from_waker(&waker);
+    loop {
+        task.state.store(POLLING, Release);
+        if let Poll::Ready(()) = unsafe { Pin::new_unchecked(&mut *task.future).poll(&mut cx) } {
+            task.state.store(COMPLETED, Release);
+        }
+        match task.state.compare_exchange(POLLING, IDLE, AcqRel, Acquire) {
+            Ok(_) => break,
+            Err(NOTIFIED) => continue,
+            _ => unreachable!(),
+        }
+    }
+}
+```
+
+Barbara finished her initial implementation of the async executor. Despite there were a lot more possible optimizations, Barbara already felt it is a bit complex. She was also confused about why she needed to care so much about polling and waking while her initial requirement was just adding additional information to the task for customizing scheduling.
+
+## 🤔 Frequently Asked Questions
+
+*Here are some standard FAQ to get you started. Feel free to add more!*
+
+* **What are the morals of the story?**
```

Comment:
The style of the FAQ sections was changed in #118 / #119, using sub headings (###) instead of nested bullet lists.